### DISCUSSION TITLE
Null check operator used on a null value when testing migrations in null safety mode

### DISCUSSION THREAD

**Discussion by MiguelSOliveira** [simolus3/drift#1637] [Q&A]
Hey!

I'm trying to migrate from moor to drift, but when I run the command I get the following analyzer error:

```
➜  git:(migrate-null-safety) flutter pub run moor_generator migrate
Failed to build moor_generator:moor_generator:
../../../.pub-cache/hosted/pub.dartlang.org/analyzer-2.8.0/lib/src/error/best_practices_verifier.dart:312:50: Error: The getter 'displayString' isn't defined for the class 'TargetKind'.
 - 'TargetKind' is from 'package:meta/meta_meta.dart' ('../../../.pub-cache/hosted/pub.dartlang.org/meta-1.3.0/lib/meta_meta.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named 'displayString'.
        var kindNames = kinds.map((kind) => kind.displayString).toList()
                                                 ^^^^^^^^^^^^^
../../../.pub-cache/hosted/pub.dartlang.org/analyzer-2.8.0/lib/src/error/best_practices_verifier.dart:314:36: Error: The getter 'commaSeparatedWithOr' isn't defined for the class 'List<dynamic>'.
 - 'List' is from 'dart:core'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'commaSeparatedWithOr'.
        var validKinds = kindNames.commaSeparatedWithOr;
                                   ^^^^^^^^^^^^^^^^^^^^
../../../.pub-cache/hosted/pub.dartlang.org/analyzer-2.8.0/lib/src/error/best_practices_verifier.dart:1702:40: Error: Member not found: 'topLevelVariable'.
      return kinds.contains(TargetKind.topLevelVariable);
                                       ^^^^^^^^^^^^^^^^
pub finished with exit code 1
```

**Comment by simolus3** [comment]
It looks like pub gave you a broken `analyzer` version. If you make sure you're running the latest version of `moor_generator` and run `flutter pub upgrade`, can you get a more recent analyzer version like 3.1.0?
You can also temporarily add `analyzer: ^3.1.0` to your `dev_dependencies` so that pub gives you an error if the latest analyzer is not available with your dependencies.

**Reply by MiguelSOliveira** [reply]
Now I get the same error with 3.1.0 :x

```
Failed to build moor_generator:moor_generator:
../../../.pub-cache/hosted/pub.dartlang.org/analyzer-3.1.0/lib/src/error/best_practices_verifier.dart:312:50: Error: The getter 'displayString' isn't defined for the class 'TargetKind'.
 - 'TargetKind' is from 'package:meta/meta_meta.dart' ('../../../.pub-cache/hosted/pub.dartlang.org/meta-1.3.0/lib/meta_meta.dart').
Try correcting the name to the name of an existing getter, or defining a getter or field named 'displayString'.
        var kindNames = kinds.map((kind) => kind.displayString).toList()
                                                 ^^^^^^^^^^^^^
../../../.pub-cache/hosted/pub.dartlang.org/analyzer-3.1.0/lib/src/error/best_practices_verifier.dart:314:36: Error: The getter 'commaSeparatedWithOr' isn't defined for the class 'List<dynamic>'.
 - 'List' is from 'dart:core'.
Try correcting the name to the name of an existing getter, or defining a getter or field named 'commaSeparatedWithOr'.
        var validKinds = kindNames.commaSeparatedWithOr;
                                   ^^^^^^^^^^^^^^^^^^^^
../../../.pub-cache/hosted/pub.dartlang.org/analyzer-3.1.0/lib/src/error/best_practices_verifier.dart:1702:40: Error: Member not found: 'topLevelVariable'.
      return kinds.contains(TargetKind.topLevelVariable);
                                       ^^^^^^^^^^^^^^^^
pub finished with exit code 1
```

**Reply by MiguelSOliveira** [reply]
i'm trying to migrate because I'm getting this error when running the migration_tests after migrating to null safety:
`Null check operator used on a null value`

This is due to me using `moor_generator: 4.2.2`, right?

**Reply by simolus3** [reply]
> Now I get the same error with 3.1.0 :x

I think the `analyzer` package is to blame here, but can you add `meta: ^1.7.0` to your pubspec too?

> Null check operator used on a null value

It could be due to the old version of `moor_generator`, but can you post the stack trace or more information to help me look into this?

**Reply by MiguelSOliveira** [reply]
Just so it's easier for me, can you help me with my second problem? 

Because I would like to avoid migrating to drift at the moment, since I'm already migrating to null safety. And then just handle migrating to drift in a future pirce of work.

(Unless this problem forces me to upgrade to drift)

----------------------


The problem is, I'm trying to migrate to null safety and when running my migration unit tests, some fail due to this error:
`  Null check operator used on a null value`

It doesn't tell me much else, this is the stack trace:
```
  package:x/database/database.g.dart 5887:56                new x.fromData
  package:x/database/database.g.dart 6322:24                x.map
  dart:_internal                                                              ListIterable.toList
  package:moor/src/runtime/query_builder/statements/select/select.dart 56:35  SimpleSelectStatement._getWithQuery
```

On the same test I get this warning:
```
WARNING (moor): It looks like you've created the database class DatabaseAtV10 multiple times. When these two databases use the same QueryExecutor, race conditions will occur and might corrupt the database.
```

I believe the problem is that, in my unit tests, I have this:
```
final schema = await verifier.schemaAt(9);
final rawDb = schema.rawDatabase;
rawDb.execute(
  'INSERT INTO ...',
);
rawDb.execute(
  'INSERT INTO ...',
);

db = MoorDb.connect(schema.connection);
....
 final x = await db.select(db.x).get();
....
```

But in my stacktrace I can see that moor goes to database.g.dart and there I see that these fields are non nullable. Ex:
`xPoints: intType.mapFromDatabaseResponse(data['${effectivePrefix}x_points'])!,`

But this `xPoints` should not exist in this test (in v9), yet.

This occurs on this line:
`final x = await db.select(db.x).get();`

I then do some expects to make sure the data was entered correctly.

Does it make sense?

**Reply by simolus3** [reply]
> But this xPoints should not exist in this test (in v9), yet.

When you do `MoorDb.connect`, are you making sure that you're also using the correct version of the database in question? Using the latest database for this purpose will not work for the reason you mentioned (it now expects columns that aren't there yet).

You can pass `--data-classes` and `--companions` to `moor_generator schema generate` to also generate snapshots of old data classes, which can be useful for this purpose.

**Reply by MiguelSOliveira** [reply]
Sorry Simon.. Still not working :X

Here's my current code:
```
final schema = await verifier.schemaAt(9);

final oldDb = v9.DatabaseAtV9.connect(schema.newConnection());
await oldDb.into(oldDb.x).insert(
      const v9.xCompanion(
        valueX: Value('x'),
      ),
    );
await oldDb.into(oldDb.y).insert(
      const v9.yCompanion(
        yName: Value('y')
      ),
    );
await oldDb.close();

final db = MoorDb.connect(schema.newConnection());
await verifier.migrateAndValidate(db, 10, validateDropped: true);
```

The problem comes when in the `migrateAndValidate`, which accesses the database.g.dart (if I comment this line, the test passes):
```
  Null check operator used on a null value
  package:x/database/database.g.dart 1234:57                new x.fromData
```

This is because the `y` table, has a `NOT NULL` field in the database.g.dart that doesn't exist yet in v9 and v10.

I can't see what I'm missing :X

**Reply by simolus3** [reply]
Are you reading rows in your migration callback of `MoorDb`? If so, you'll probably have to rewrite your migration logic to be aware of this problem. The verifier should not read rows by default, can you post the full stack trace?

There's an example for migration tests (including a migration where we add a non-nullable column) [here](https://github.com/simolus3/moor/tree/develop/extras/migrations_example). If you don't want to share too many details of your project (which is fine), perhaps that example helps to find potential differences.

**Reply by MiguelSOliveira** [reply]
I am following the `migration_test` you mentioned, yeah.

`Are you reading rows in your migration callback of MoorDb?`
I am, yes. You do as well, in the migration_test file, right?
```
    final migratedDb = v2.DatabaseAtV2.connect(schema.newConnection());
    final user = await migratedDb.select(migratedDb.users).getSingle();
```
`The verifier should not read rows by default`
Can you clarify? Because I'm using the verifier to migrateAndValidate, not to read rows, right?

`can you post the full stack trace`
Unfortunately, what I've posted above is the full stacktrace I get:
```
  package:x/database/database.g.dart 5887:56                new x.fromData
  package:x/database/database.g.dart 6322:24                x.map
  dart:_internal                                                              ListIterable.toList
  package:moor/src/runtime/query_builder/statements/select/select.dart 56:35  SimpleSelectStatement._getWithQuery
```

**Reply by simolus3** [reply]
> You do as well, in the migration_test file, right?

We do it in the test file,  but we're using the `DatabaseAtV2` to read rows - and those row classes at version two don't expect all the new columns. We don't read any rows in the migration callback [here](https://github.com/simolus3/moor/blob/2a347b2f40702df7f1679c5b9cc03b8d72ada079/extras/migrations_example/lib/database.dart#L15-L45).

> Can you clarify? Because I'm using the verifier to migrateAndValidate, not to read rows, right?

Exactly! `migrateAndValidate` will call the `MigrationStrategy` on your database and read the schema of your database for a diff. It doesn't read contents of your database.

**Reply by MiguelSOliveira** [reply]
This is my migration logic. It is reading rows with the MoorDb param. Is that the source of the problem? How can i re-write it?

```
Future<void> migrate9(Migrator migrator, MoorDb db) async {
  // ignore: avoid_print
  print('Migrating v9');
  final newDb = v10.DatabaseAtV10(_stub);
  await migrator.createTable(newDb.x);
  await migrator.createTable(newDb.y);

  final z = await db.select(db.z).getSingle();
  final zz = ...;

......
```

**Reply by simolus3** [reply]
If you migrate from a version where `z` did not have a column that it has now, then yes, that is the problem. You can use the `--data-classes` option on `moor_generator schema generate` to also generate old versions of data classes and then use `newDb.select(newDb.z)` to obtain the right version of the data class.

**Reply by MiguelSOliveira** [reply]
do i also add entries to the new db instead of Moordb? I've already generated with "--data-classes" so I think I'm good there.

Like this:
```
....
await newDb.select(newDb.x).getSingle();
....
await newDb.into(newDb.y).insert(....
.....
```

Where newDb is `v10.DatabaseAtV10(_stub);`?

Basically don't use MoorDb for the migration?



Also this line "await newDb.select(newDb.x).getSingle();" returns a DataClass. How can I make it return the correct type of newDb.x?

**Reply by simolus3** [reply]
> Also this line "await newDb.select(newDb.x).getSingle();" returns a DataClass. How can I make it return the correct type of newDb.x?

It should already? The generated table from `newDb` will also map to a data class from v10 if you run it with `--data-classes`.

> do i also add entries to the new db instead of Moordb

That's not always necessary, but probably the safest way to avoid columns that shouldn't be there yet.

**Reply by MiguelSOliveira** [reply]
> It should already? The generated table from `newDb` will also map to a data class from v10 if you run it with `--data-classes`.

You mean it should already return a DataClass? Sorry, I didn't understand. I need it to return an instance of X which is the type of the table. Does it make sense?

~~It seems to work with casting "as X". Is that the right way to do it?~~

EDIT: The above does not seem to work:
```type 'XsData' is not a subtype of type 'X in type cast```

**Reply by MiguelSOliveira** [reply]
Sorry bro, I'm really confused... Maybe I'm using migration testing incorrectly? This doesn't seem to be a common problem looking at existing issues, but I think it's the best way to test migrations... Maybe I'm misunderstanding drift.

**Reply by simolus3** [reply]
Maybe it would be easier to see what's going wrong if you can create a simple repository where the same error happens? Then it's much easier for me to take a look and tell you where exactly it's going wrong :) All of this `XsData` and `X` obfuscation doesn't make it obvious what you're doing haha.
